platform/default/include/mbgl/util/default_thread_pool.hpp

```diff
@@ -6,17 +6,40 @@
 #include <mutex>
 #include <queue>
 #include <thread>
+#include <functional>
 
 namespace mbgl {
 
+struct ThreadLifecycle {
+    using ThreadData = void*;
+    ThreadLifecycle() = default;
+    ThreadLifecycle(std::function<ThreadData()> onThreadCreated_,
+                    std::function<void(ThreadData)> onThreadDestroyed_) :
+            onThreadCreatedFn(onThreadCreated_),
+            onThreadDestroyedFn(onThreadDestroyed_) {}
+
+    ThreadData onThreadCreated() const {
+        return onThreadCreatedFn();
+    }
+
+    void onThreadDestroyed(ThreadData threadData_) const {
+        onThreadDestroyedFn(threadData_);
```

Comment:
If `threadData_` would be private, we could null it after `onThreadDestroyedFn` and assert if someone calls `ThreadLifecycle::onThreadDestroyed` twice.

Reply:
We cannot assert if it's not null because the default impl returns null.